Add try-catch for GLTFBinaryExtension

takahirox opened the pull request:
I think here would be better to throw error rather than using `onError` callback

https://github.com/mrdoob/three.js/blob/47e792705d29e547660d4a5ef9fd1bff6a41e443/examples/js/loaders/GLTFLoader.js#L94-L99

for the consistency (`GLTFBinaryExtension` throws error).

And `GLTFLoader.parse` is public method. It'd be better to throw for error synchronously occurred.

Even if we use `onError`, we need `if ( onError !== undefined )` check because `onError` is optional.

/cc @donmccurdy

takahirox committed:
Replace callback with throwing error for unsupported glTF version in GLTFLoader

donmccurdy reviewed (approved):
I don't have a strong preference on this one... it seems like if we're treating `parse()` as a public method and the user passes in an `onError` argument we'd ideally always respect that (even though GLTFBinaryExtension doesn't...). But wrapping everything in try/catch isn't pretty either.

takahirox commented:
TBH, I don't have strong preference neither. But I just wanted consistency.

But current code, even with this change, still has another inconsistency in `.parse`. Some errors are handled via `onError` while others are thrown even if `onError` is passed in.

The tricky part of `.parse` is it can invoke sync and async errors. try-catch can't handle async errors.

So, letting `onError` be requirement and always using it, not throwing, would be another option. It'd be simpler because it can handle both sync/async errors.

And `GLTFBinaryExtension` errors can be handled like this

    try {

        extensions[ EXTENSIONS.KHR_BINARY_GLTF ] = new GLTFBinaryExtension( data );

    } catch( error ) {

        onError( error );

    }

donmccurdy commented:
I'm happy with that, sure.

takahirox committed:
Revert "Replace callback with throwing error for unsupported glTF version in GLTFLoader"

This reverts commit bbfca2a009a0e16ab4d3306265369304ee832d70.

takahirox committed:
Add try-catch for GLTFBinaryExtension

takahirox committed:
Merge branch 'dev' into UpdateGLTFLoader

takahirox committed:
Update GLTFLoader documentation

takahirox commented:
Updated.

I added try-catch for `GLTFBinaryExtension`. `onLoad` can handle both sync and async errors now.
And I removed "(optional)" for `onLoad` and `onError` from `GLTFLoader` documentation.

donmccurdy commented on docs/examples/loaders/GLTFLoader.html:
```diff
@@ -125,7 +125,7 @@
 		[page:String path] — The base path from which to find subsequent glTF resources such as textures and .bin data files.<br />
 		[page:Function onLoad] — A function to be called when parse completes.<br />
-		[page:Function onError] — (optional) A function to be called if an error occurs during parsing. The function receives error as an argument.<br />
+		[page:Function onError] — A function to be called if an error occurs during parsing. The function receives error as an argument.<br />
```
I think it would be better to keep both `onError` callbacks optional. Can't have a required parameter after an optional one (`onProgress`) in any case.

takahirox replied:
So, replacing 

    onError( error );

with

    if ( onError ) onError( error );

?

donmccurdy replied:
👍

takahirox committed:
Add if ( onError ) check in GLTFLoader

takahirox committed:
Add (optional) for onError in GLTFLoader documentation

takahirox commented:
Updated.

Note that if `onError` isn't passed in from users, some sync errors aren't notified to them and other errors invoked in Promises are uncaught and thrown to them.

donmccurdy reviewed (approved):
Looks consistent to me, thanks!

mrdoob commented:
Thanks!
